### ISSUE TITLE
Unable to export the tickets

### ISSUE THREAD

**Issue by umisuhanarahman** [osTicket/osTicket#5303]
![exportxerror](https://user-images.githubusercontent.com/33883955/71404028-6a4d5d80-266c-11ea-99f8-fae8b4e78bba.png)

Server Information
osTicket Version	v1.14.1 (f1e9e88) —  Up to date
Web Server Software	Apache
MySQL Version	10.2.29
PHP Version	7.3.12

Hello, can you assist me how to solve this. I cannot export my tickets to csv because when I click on the export button the pop up goes blank just like the picture above. What I need to do to able to export the ticket csv file?

My console show this "Failed to load resource: the server responded ajax.php/export/10kUeB/check:1 with a status of 404 () "

Plus, I checked that there is no error in any logs.

![permission](https://user-images.githubusercontent.com/33883955/71404299-260e8d00-266d-11ea-88c3-3b17b7e8eae2.png)

I also check all the permission for class.export.php file.

On the other hand, I already tried use the export function in the v1.12.5 which is the older version and the export function run smoothly but it won't do the same in this latest version.

I hope anyone could help me to solve this. Thank you in advance!

**Comment by JediKev** [comment]
@umisuhanarahman 

Check the permissions of the temp directory. PHP will create a file in the temp directory to write the data and then reads from the file to send the data to the browser for download or to email the file to you. Maybe PHP is struggling to read from the file or is struggling to check the status of the export?

Cheers.

**Comment by gtao725** [comment]
I have the exact same issue but a little different, for me this only happens from Agent Panel->Users->Organizations->Choose Any Organization->Tickets->Export. I am also on 1.14.1 upgraded from 1.12.5 (it worked fine in that version). Is there a fix or temporary workaround to this? It is the end of the month and we need to generate customer billing info based on per-organization basis. Pls kindly help! Thanks in advance!

**Comment by umisuhanarahman** [comment]
@JediKev 

![permission](https://user-images.githubusercontent.com/33883955/71460216-eb7c3000-27e5-11ea-99da-b98fc4fa285d.png)


Is this the correct temp folder that you meant? I already gave all permissions 
but the problem isn't solve.

If I'm wrong, where can I find the temp directory?

Thanks.

**Comment by ntozier** [comment]
No one here can tell you where your temp folder is located. You would have to consult your php.ini to locate your temp directory.  

If you do not know where your php.ini is located, then you can create a file in your website root folder (give it any name you want, like delme.php) and in it put
```<?php phpinfo(); ?>```

Then browser to it with with your web browser.  It should tell you where your php.ini is located as well as where your temp folder is located.

**Comment by JediKev** [comment]
@gtao725 

Do you see any errors in your Apache error logs, PHP error logs, browser console, or osTicket System Logs? Also, have you checked your permissions for PHP’s temp directory?

Cheers.

**Comment by gtao725** [comment]
@JediKev 

Answers to your questions:

1. osTicket System Logs - all good, nothing out of ordinary, error level set to DEBUG so I see everything.
2. PHP Temp Dir for me is set to: sys_temp_dir = "/tmp", permission for /tmp is 777 and owned by root.
3. Apache log for the per organization ticket export operation returns this error (500 error on the very last export operation. The operations preceeding the very last returns 200 so they are ok.): 

210.22.112.50 - - [27/Dec/2019:11:03:28 +0800] "GET /scp/orgs.php?_pjax=%23pjax-container HTTP/1.1" 200 12174 "https://support.jetservices.com.cn/scp/users.php" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36"
210.22.112.50 - - [27/Dec/2019:11:03:29 +0800] "GET /scp/orgs.php?id=2&_pjax=%23pjax-container HTTP/1.1" 200 9202 "https://support.jetservices.com.cn/scp/orgs.php" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36"
210.22.112.50 - - [27/Dec/2019:11:03:34 +0800] "GET /scp/orgs.php?_pjax=%23pjax-container HTTP/1.1" 200 12174 "https://support.jetservices.com.cn/scp/orgs.php?id=2" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36"
210.22.112.50 - - [27/Dec/2019:11:03:35 +0800] "GET /scp/orgs.php?id=4&_pjax=%23pjax-container HTTP/1.1" 200 19186 "https://support.jetservices.com.cn/scp/orgs.php" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36"
210.22.112.50 - - [27/Dec/2019:11:03:38 +0800] "GET /scp/ajax.php/orgs/4/tickets/export HTTP/1.1" 500 - "https://support.jetservices.com.cn/scp/orgs.php?id=4" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36"

4. PHP not generating any errors as nothing is written to my /var/log/php.log (custom php error log and error logging is turned on in php).

So I am very lost as to what is causing this. Pls help!

Cheers!

**Comment by umisuhanarahman** [comment]
@JediKev @ntozier 

Thank you for your help to find temp directory. I already found the path in php.ini file as below 

![php ini](https://user-images.githubusercontent.com/33883955/71502166-838f1d80-28a9-11ea-9f52-8283ad80214b.png)

I had given all the permission to all the temp directories as stated in the path "/var/cpanel/php/sessions/ea-php72"  and I had set all the file and directories as default permissions, but the problem still doesn't solve.

What else can I do next?

Thanks.

**Comment by ntozier** [comment]
Your screen shot does not appear to show the temp directory... The temporary directory directive is `upload_tmp_dir`.  session.save_path is the directory where PHP stores session details.

**Comment by gtao725** [comment]
Also set my upload_tmp_dir = "/tmp", rebooted server as well. Double checked persmission for /tmp and it still is the 777 permission but the problem STILL exists unfortunately. I seriously do not think it's the permissions since a previous version of osTicket worked flawlessly.......

**Comment by JediKev** [comment]
@gtao725 

The exporter was completely changed in 1.14.x. In 1.12.x and below the system builds the export and pushes the data directly to the browser for download all in the same request which was obviously not too great as it caused timeouts on large exports. In 1.14.x the system builds the export in a temp file in the background and makes repeated AJAX requests every X seconds to check the status of the export, if the export takes a long time the system sends the export in an email to the address associated with your Agent account. It also has a button on the "Please Wait" window to email the export if you don’t feel like waiting.

With all this being said, the issue seems to stem from when it checks the status of the export (at least in the cases we have been given debug results). When it checks the status it first checks the session for an export by a particular ID (each export has its own unique ID in case you have multiple exports going on) and a file within that particular export. If it has both of those things it tries to create an Exporter object and return it to the subsequent methods. So it might be the system having issues saving the export object in the session, or maybe having issues constructing an Export object in PHP, or having issues reading the temp file, or having issues locking/checking the lock on the temp file, etc. We can’t really tell yet as we can’t replicate the issue in house. If you or anyone can provide any helpful debug information we can hopefully get this resolved.

I’d start by going to `include/class.export.php` to the `Exporter::load()` function (around line 556) and do a `var_dump()` of the session to see if it has the export object and the file within it.

Cheers.

**Comment by gtao725** [comment]
ok thx, I will try turn on debug info inside php.ini as well to see if it throws anything onscreen.,

**Comment by umisuhanarahman** [comment]
@ntozier @JediKev 

I'm sorry for the wrong file but I already found the temp directory which is  '/tmp' and it had all the permissions = 777 but the problem doesn't solve too. For the debug info you said earlier, I did the var_dump and get the results as below: 

![var_dump](https://user-images.githubusercontent.com/33883955/71567238-7451df00-2af8-11ea-8e20-8b1c91f83525.png)

From the screenshot above, I think the export object is already there right? 
So how is it?

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

So, for your case you can see `No such export` is being returned which the check responsible for this is [here](https://github.com/osTicket/osTicket/blob/v1.14.1/include/ajax.export.php#L11-L12). The issue is either the system cannot load the Exporter instance (via `Exporter::load($id)`) or the Exporter is not available (via `$exporter->isAvailable()` which means PHP's built-in `file_exists()` method fails when checking the temp file; in the case above it was `/tmp/24jgxxxx`).

You can see if the Exporter instance is being loaded by putting `var_dump($exporter);` in `include/class.export.php` on line `576`. If you get an Exporter instance of `CsvExporter` then the exporter is created and you know it's failing on PHP's `file_exists()` method.

Cheers.

**Edit:**

If `file_exists()` is causing issues (not `Exporter::__construct` or `Exporter::getFile()`) I'm curious if putting the following in `include/class.export.php` on line `396` would solve the issue?
```
     function isAvailable() {
         clearstatcache();
```

If you are having issues with the `file_exists()` method, try the above and let me know if it resolves the issue.

**Comment by umisuhanarahman** [comment]
@JediKev 

Thanks for the reply sir. I did what you said to put **var_dump($exporter);** in **include/class.export.php** on line **576** but nothing appear on the screen. Then, I try place the code on the other line which is on line **565** and this is the output

![565](https://user-images.githubusercontent.com/33883955/71609575-9dd83c80-2bc4-11ea-8188-a98b0e8e4f10.png)

Plus, I also tried to put on line **588** just to make sure is it true that the exporter is **NULL** and this is the output on screen.

![588](https://user-images.githubusercontent.com/33883955/71609664-3e2e6100-2bc5-11ea-824f-f11f7ba5060c.png)

So, both show that the exporter was NULL. I think the first possibility that you mention is the issue which is **the system cannot load the Exporter instance (via Exporter::load($id))**. But, this is just my thought, please correct me if I'm wrong. I'm not sure that the var_dump($exporter) output is true because I put in other line. If it is true, what I need to do to fix the issue? In your previous reply you give the solution for the second possibility only. 

Besides, I had tried your code for
**If file_exists() is causing issues (not Exporter::__construct or Exporter::getFile()) I'm curious if putting the following in include/class.export.php on line 396 would solve the issue?**

     function isAvailable() {
         clearstatcache();

I try it just to see if something happen but nothing different thing appears.

So, what can I do next, sir? I'm looking forward for your reply and really appreciate your help.

Happy New Year and Thanks.

**Comment by umisuhanarahman** [comment]
@JediKev @ntozier 

Hello sir, I want to follow up about my previous reply. So how the solution I could do to solve this issue. I hope you can help me. Your reply means a lot.

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

Now you should go to the `Exporter::load()` function again and this time put `var_dump($id, $info, $class);` on line 572 and put `var_dump($ex);` after `} catch (Exception $ex) {`. The first will confirm you have an ID, the info, and the class to instantiate. The second will show you an Exception if there was an issue creating the new instance. Please post both results here.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir, 

I did what you told me to do. And this is the output:

**a) var_dump($id, $info, $class); on line 572**

![var_dump($id, $info, $class);](https://user-images.githubusercontent.com/33883955/71711994-65439780-2e3e-11ea-9195-0f5cc038f3d7.png)

**b) put var_dump($ex); after } catch (Exception $ex) {**

![var_dump($ex)](https://user-images.githubusercontent.com/33883955/71712025-7e4c4880-2e3e-11ea-8fde-313fa7ef1b02.png)

Why both of them got almost the same output and only the second parameter was different and change in every try? What can you conclude from this output?

Thanks.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir.

Regarding to my previous reply, what can I do next sir? Looking forward to hear your idea of solution to fix this issue!

Thanks.

**Comment by umisuhanarahman** [comment]
@JediKev @ntozier 

Hello sir, is it there is no other solution that you could give to me? Does the labels mean there is nothing you could help to solve this issue? I really need to use this function. Looking forward for any solution from osTicket team on this issue. I hope osTicket team could debug this problem and release the new version that fix those issues because I realized that's not just me who got this problem. There's many people facing the same problem on the latest version of osTicket v1.14.1 that you can see from discussion in osTicket forum.  Appreciate to hear your updates soon.

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

>Hello sir, is it there is no other solution that you could give to me? Does the labels mean there is nothing you could help to solve this issue?

The labels just mean we do not know the cause and we need to review this again. We cannot replicate this issue in house so it's very difficult to make a patch. We will continue to assist you in debugging the issue as best as we can.

>Regarding to my previous reply, what can I do next sir?

First, remove all previous `var_dump` statements as these just cloud the current dumps. Next, you can go to the `Exporter::load()` function and put:
```
$info = $_SESSION['Exports'][$id];
$class = $info['class'];
$exporter = new $class(array('id' => $id) + $info);
var_dump('ID: ', $id, 'Info: ', $info, 'Class: ', $class);
var_dump('Exporter: ', $exporter);
```
.. **before/above** `try {`. Once the changes are made and saved, try to export again and post the results back here.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

I did remove all the previous var_dump statements and put those code in the Exporter : : load():

![code](https://user-images.githubusercontent.com/33883955/71945428-c0400a80-3201-11ea-8556-bb5dd5e8c4c2.png)

But, it doesn't solve the issue and the output was still the same:

![again](https://user-images.githubusercontent.com/33883955/71945476-e5cd1400-3201-11ea-8fac-4a647b4e8f4e.png)

So, how is it sir?

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

Of course it’s not going to address the issue, this is just the debugging stage. We have to figure out what the issue is before we can fix it.

Above `if (!isset($_SESSION['Exports'][$id])` put:
```
var_dump('ID: ', $_SESSION['Exports'][$id], 'File Exists: ', file_exists($_SESSION['Exports'][$id]['file']));
```

I have a feeling your PHP is having issues reading the file.

**Note:**
Tbh, I should’ve done this in reverse but I assumed and jumped the gun. My apologies.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

Okay, after I put those code, this is the output:

![next](https://user-images.githubusercontent.com/33883955/71947837-0f3d6e00-3209-11ea-90cc-af9222897916.png)

What means by that var_dump output?

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

As I suspected your PHP is having issues reading the file as `file_exists` is failing. Ensure PHP has the appropriate permissions on the file in the Temp directory. If you don’t know how refer to the many guides/documentation available online. Also, you need to make sure the files are actually being created by looking at the temp directory during export and ensuring a file is being created.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

But, as i told you before. I already gave all 777 for the permission to the tmp files and folders as below (this is the /tmp directory):

![tmp](https://user-images.githubusercontent.com/33883955/71949068-19616b80-320d-11ea-8080-e69f9d0d236e.png)

What other permission that I need to give? Did I gave the wrong permission? 

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

As per PHP documentation:
> Warning
This function returns FALSE for files inaccessible due to safe mode restrictions. However these files still can be included if they are located in safe_mode_include_dir.

So I’d start by checking the safe mode restrictions to ensure this is not the culprit.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

How can I check the safe mode restriction, sir?

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

You’ll have to read PHP's documentation on this available on their website.

You can also try adding `clearstatcache();` right below `static function load($id) {` to see if that helps.

Cheers.

**Comment by umisuhanarahman** [comment]
> You can also try adding clearstatcache(); right below static function load($id) { to see if that helps.

Nothing change on the output:

![noob](https://user-images.githubusercontent.com/33883955/71949704-870e9700-320f-11ea-807a-c531ffde8f23.png)

> You’ll have to read PHP's documentation on this available on their website.

I will read it first.

Thanks.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir,

Regarding to previous reply about

> So I’d start by checking the safe mode restrictions to ensure this is not the culprit.

I already read the PHP documentation about the safe mode that you mentioned. And I had checked that the safe mode is default which is `safe_mode = Off` and from the documentation that 

> However these files still can be included if they are located in safe_mode_include_dir.

I tried to change the setting for `safe_mode_include_dir = /tmp` if in case it could solve the issue but I'm using the shared hosting so that I can't proceed the changes I made because my web hosting didn't allow it

> We would like to inform you that we cannot enable safe_mode due to it would require to be done on a server level which we cannot afford due to this being a shared hosting environment and we cannot afford to affect all the users.

and they also said

> Regarding the permissions of the /tmp folder, you can see that the permissions are 777 which you can see via the File Manager tool. Considering this, these are the full permissions and there should be no issue regarding that.

So how is it sir? What else I can do to solve this issue? Glad to hear your reply!

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

Did you do what I mentioned previously?
>Also, you need to make sure the files are actually being created by looking at the temp directory during export and ensuring a file is being created.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir, 

I already did what you mentioned but I cannot find that file was created in the temp directory. I think it's not created. How it is sir?

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

As you can see here:
- https://github.com/osTicket/osTicket/blob/develop-next/include/class.export.php#L403-L430

We attempt to create a file in the temp directory, and if we can’t we throw an Exception which you should see in your server error logs or PHP error logs. Go to this function and put:
```
$prefix = Format::filename(Misc::randCode(6));
$temp=tempnam(sys_get_temp_dir(), $prefix);
$this->fp=fopen($temp, $mode);
$meta=stream_get_meta_data($this->fp);
var_dump('prefix: ', $prefix, 'temp: ', $temp, 'fp: ', $this->fp, 'meta: ', $meta);
die;
```

.. above the `try {` statement to see if all variables are being set.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

Thank you for your previous reply.

So this is the output from those codes:

![only exception](https://user-images.githubusercontent.com/33883955/72120141-e2689280-3391-11ea-897b-6fa68b6ad02d.png)

What does it mean sir?

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

It looks like everything is there...I don’t know why `file_exisits` is failing for you.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

So what else can I do to fix this sir?

Thanks

**Comment by JediKev** [comment]
@umisuhanarahman 

At this point I’m not sure. I’ll have to do some research.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

I'm looking forward for your solution sir. Please get back to me when you have anything that can help. I'm really appreciate your reply.

Thanks.

**Comment by mhakim168** [comment]
I got the same problem, I have an account that cant export the ticket, but failed export using another account.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir, 

So how about this issue? Did you find any solution to solve this or did you had figured out the cause of this issue? I'm waiting for your reply. Please get back to me soon.

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

Okay, so on your server, find Apache's systemd service file (it's different for every server so please google where to find this file on your specific server), set `PrivateTmp` to `false`, save the file, restart Apache, and retest the Export.

IF the above does not work for you then try the following:
- Login to your server
- Go to osTicket root folder (where you see `api/`, `include/`, etc.)
- Create a new folder called `tmp` inside osTicket root folder
- Give this folder `0777` permissions (for testing purposes)
- Go to `include/class.export.php` line `414`
- Change
```
if (!($temp=tempnam(sys_get_temp_dir(), $prefix))
```
to
```
if (!($temp=tempnam(ROOT_DIR . 'tmp/', $prefix))
```
- Retest the export

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir, 

Thank you for your reply.

About what you said earlier

> find Apache's systemd service file (it's different for every server so please google where to find this file on your specific server), set PrivateTmp to false, save the file, restart Apache, and retest the Export.

I can't access to Apache systemd service file because I'm using shared hosting. I already asked my web hosting and they said we didn't allowed to access those file..

Then, I'd try the steps you said in the previous reply:

> Login to your server
> Go to osTicket root folder (where you see api/, include/, etc.)
> Create a new folder called tmp inside osTicket root folder
> Give this folder 0777 permissions (for testing purposes)
> Go to include/class.export.php line 414
> Change
> if (!($temp=tempnam(sys_get_temp_dir(), $prefix))
> to
> 
> if (!($temp=tempnam(ROOT_DIR . 'tmp/', $prefix))
> Retest the export

I had done all from the above but the issue doesn't solved yet, the export still not functioning but the file was created in the /tmp file, however the file is empty. 

![tmpfile](https://user-images.githubusercontent.com/33883955/72495848-d3d31d00-3862-11ea-954a-1d3885f14409.png)

Then I tried to put back the previous var_dump that you told me and this is the output:

![again](https://user-images.githubusercontent.com/33883955/72495274-388d7800-3861-11ea-89b4-9d844a844cb5.png)


So how is it sir? Is there any solution that you can give me to do next? I really looking forward for your reply.

Thanks.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir, 

Regarding my previous reply, what you can conclude from it? What can I do next to solve this issue? I really appreciate your reply.

Thanks

**Comment by JediKev** [comment]
@umisuhanarahman 

We are looking further into the issue and will update this thread when we find something.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

Thank you for your assistance. I really looking forward for your updates. Will wait for your reply soon!

Thanks

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir.

I would like to follow up this issue. Did you had figured out any idea how to solve it? I trust you are doing your best in solving this problem. I'm looking forward for your upcoming reply regarding this issue. Your help means a lot to me. 

Thanks.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir, 

I have an update regarding to this issue which is I just realized that I got the export file using the email but not auto downloaded when export button clicked. Is it really there's a new features in this 1.14.1v that the export file will send to the email registered in the My Profile? It is because I've tried using 1.12.5v there's no export file send to the email but it auto downloaded when the export button clicked. Below is the screenshot of email sent:

![emailexport](https://user-images.githubusercontent.com/33883955/73511911-77572c80-4422-11ea-9d8e-202506182318.png)

I just want to ask how is it happening? Is it really there is auto email for the export file? And why the screen remain the same as below while the file already sent through email?

![empty](https://user-images.githubusercontent.com/33883955/73512049-1845e780-4423-11ea-98d1-5fe6a7becb3f.png)

I hope this update will brighten up the solution for this issue. Looking forward for your idea of solution to solve this issue!

Thanks.

**Comment by umisuhanarahman** [comment]
Hello sir, 

How's the update about this issue? It's been awhile since the last time I got the reply from you. Did you have anything yet to help me solve this? I'm looking forward for your respond regarding this issue. If you still have nothing new solution, I hope you could reply me with some follow up or else. Just to make sure that you still reviewing on this issue or not.

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

The update is that we have setup an instance on a hosting platform that is experiencing the same issue. We are now able to replicate the issue and are looking into the cause. Any updates will be posted here. If there are no posts here then there are no updates and you should wait for an update to be posted.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir, 

Thank you for the reply! Noted that and will be looking forward for your updates soon. I really appreciate your respond sir.

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

Okay, so you are on a shared host right? It's most likely a shared host configuration to disable output buffering. In my instance of shared hosting, my host told me that they intentionally disable output buffering and that I had to upgrade to VPS or Dedicated server in order for it to work correctly. If I were you I would contact your host to see how you can enable output buffering for PHP.

Cheers.

**Comment by JediKev** [comment]
@gtao725 @mhakim168 Sorry, I forgot to reference you guys in my message above ^

**Comment by gtao725** [comment]
Thx JediKev,

I am on a dedicated VPS, could you pls tell me which PHP paramenter to
check this output buffering?

Sincerely,

Geoff Tao

**Comment by JediKev** [comment]
@gtao725 

You should ask your host as it varies. If all else fails check PHP’s output_buffering setting and make sure zdlib.output_compression is Off.

Cheers.

**Comment by gtao725** [comment]
Ok I'll check, thx!

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir, 

Thank you for your update.

I already ask my host to enable output buffering for my website and this is the reply from them
![tmd](https://user-images.githubusercontent.com/33883955/74422104-03d50680-4e89-11ea-86ac-46c312ee2de3.png)

And I also already confirm that zdlib.output_compression is Off
![image](https://user-images.githubusercontent.com/33883955/74422658-e6546c80-4e89-11ea-8989-b71f9c405484.png)


But after that I try again to export the file but it turn out to be the same issue again. How can I do this sir? Is there any other setting that I could ask my host to change? I'm looking forward for more updates from you

Thanks.

**Comment by scarcamo-eyg** [comment]
> @umisuhanarahman
> 
> Okay, so on your server, find Apache's systemd service file (it's different for every server so please google where to find this file on your specific server), set `PrivateTmp` to `false`, save the file, restart Apache, and retest the Export.
> 
> IF the above does not work for you then try the following:
> 
> * Login to your server
> * Go to osTicket root folder (where you see `api/`, `include/`, etc.)
> * Create a new folder called `tmp` inside osTicket root folder
> * Give this folder `0777` permissions (for testing purposes)
> * Go to `include/class.export.php` line `414`
> * Change
> 
> ```
> if (!($temp=tempnam(sys_get_temp_dir(), $prefix))
> ```
> 
> to
> 
> ```
> if (!($temp=tempnam(ROOT_DIR . 'tmp/', $prefix))
> ```
> 
> * Retest the export
> 
> Cheers.

Hi there.
I've modified this and now it is working partially from Ticket View and then export button, I received the email attachment but did not see the direct download.

From Organizations -> Users -> Tickes -> export , it is not working

I'm using godaddy shared hosting, osticket 1.14.1, php 7.3

Thank you all.

**Comment by JediKev** [comment]
@scarcamo-eyg 

Did you ask your host if output_buffering is On, zdlib.output_compression is Off, and nothing is blocking the requests?

Cheers.

**Comment by JediKev** [comment]
@umisuhanarahman @scarcamo-eyg @gtao725 

Do me a favor, comment out `include/ajax.tickets.php` lines `1911` and `1913` to:
```
//$exporter->email($thisstaff);
```
and
```
//@$exporter->delete();
```

... if you make those changes and the download works your issue is due to output buffering. If this is the case, then you should ask your host why the webserver is holding on to the response instead of allowing PHP to flush the response to the browser immediately using `ob_flush()`.

Cheers.

**Comment by umisuhanarahman** [comment]
@JediKev 

Hello sir,

Thank you for your reply!

> Do me a favor, comment out include/ajax.tickets.php lines 1911 and 1913 to:
> 
> //$exporter->email($thisstaff);
> and
> 
> //@$exporter->delete();

I did like what you said and the export file can be download successfully! But the email doesn't send because of the comment out.

And for this one:

>  then you should ask your host why the webserver is holding on to the response instead of allowing PHP to flush the response to the browser immediately using ob_flush()

I already ask my host and below is their reply:
![tmdreply1](https://user-images.githubusercontent.com/33883955/74515145-f54e2400-4f48-11ea-87e9-2e53ef8b8c28.png)
![tmdreply2](https://user-images.githubusercontent.com/33883955/74515148-f67f5100-4f48-11ea-8a59-fdc589a2e484.png)

So, from the code given by the host can I know where to put in the coding? Glad to hear your reply, it looks like this issue nearly to solve.

Thanks sir.

**Comment by JediKev** [comment]
@umisuhanarahman 

LOL they stole that code snippet from someone's blog (binarytides[dot]com)...anyways, the only thing we were missing was the `echo whitespace` part. It appears your server is waiting for `x` amount of characters to be returned before flushing the response.

You can address this for now by going to `include/class.http.php` and changing lines `61-63` from:
```
        // Flush the request buffer
        while(@ob_end_flush());
        flush();
```
to:
```
        // Flush the request buffer
        while(@ob_end_flush());
        for ($i = 0; $i < 1000; $i++) {
            echo ' ';
        }
        flush();
```

Once you make those changes your download should work again. With this being said, I will look into something more permanent and post an update here.

Cheers.

**Comment by umisuhanarahman** [comment]
Hello sir, 

Thank you for the reply.

> You can address this for now by going to include/class.http.php and changing lines 61-63 from:

>         // Flush the request buffer
>         while(@ob_end_flush());
>         flush();
> to:
> 
>         // Flush the request buffer
>         while(@ob_end_flush());
>         for ($i = 0; $i < 1000; $i++) {
>             echo ' ';
>         }
>         flush();

I already made the changes like you told. The direct download already perfect. But may I know how about that email and delete function that you ask me to comment out? 

> Do me a favor, comment out include/ajax.tickets.php lines 1911 and 1913 to:

> //$exporter->email($thisstaff);
> and
> 
> //@$exporter->delete();

Did I need to uncomment it? It is because I already uncomment those thing to make the email for export file being able to function again but it turns out to be blank again. So for now I still let those things comments out until I get any reply from you.

Thanks.

**Comment by JediKev** [comment]
@umisuhanarahman 

Hmmm...so let's recap here.

1.) I'm able to replicate your issue with a certain shared hosting platform.
2.) All I have to do to get Exports to work correctly is to echo the whitespaces, I don't have any other changes to the codebase.

Maybe your situation is slightly different? Maybe you actually need the implicit flush as well as the custom headers? Let's try this, remove all changes you've made so far, then you can go to `include/class.http.php` and changes the entire `flush()` method to:
```
    function flush($code, $content, $contentType='text/html', $charset='UTF-8') {
        self::response($code, null, $contentType, $charset);
        header('Cache-Control: no-cache, must-revalidate');
        header('Content-Length: '.strlen($content)."\r\n\r\n");

        // Disable buffering and compression - shared hosts
        ini_set('output_buffering', 'Off');
        ini_set('zlib.output_compression', false);

        print $content;
        // Flush the request buffer
        while(@ob_end_flush());

        // Implicit flush - shared hosts
        ini_set('implicit_flush', true);
        ob_implicit_flush(true);
        // Set content type and disable caching - shared hosts
        header('Content-Type: text/plain');
        header('Cache-Control: no-cache');
        // Echo whitespaces to flush request - shared hosts
        for ($i = 0; $i < 1000; $i++) {
            echo ' ';
        }

        flush();
        // Terminate the request
        if (function_exists('fastcgi_finish_request'))
            fastcgi_finish_request();
    }
```

**NOTE:**
It is important that you remove all previous changes so we are working with a clean slate each time we try something new.

**EDIT:**
Also, if that doesn't work, try setting `ini_set('output_buffering', 'Off');` to `ini_set('output_buffering', 'On');`.

Cheers.

**Comment by umisuhanarahman** [comment]
Hello sir, 

Thank you for the reply.

I had changed the flush() method and remove all the previous changes as you mention before

>  Let's try this, remove all changes you've made so far, then you can go to include/class.http.php and changes the entire flush() method to:
> 
>     function flush($code, $content, $contentType='text/html', $charset='UTF-8') {
>         self::response($code, null, $contentType, $charset);
>         header('Cache-Control: no-cache, must-revalidate');
>         header('Content-Length: '.strlen($content)."\r\n\r\n");
> 
>         // Disable buffering and compression - shared hosts
>         ini_set('output_buffering', 'Off');
>         ini_set('zlib.output_compression', false);
> 
>         print $content;
>         // Flush the request buffer
>         while(@ob_end_flush());
> 
>         // Implicit flush - shared hosts
>         ini_set('implicit_flush', true);
>         ob_implicit_flush(true);
>         // Set content type and disable caching - shared hosts
>         header('Content-Type: text/plain');
>         header('Cache-Control: no-cache');
>         // Echo whitespaces to flush request - shared hosts
>         for ($i = 0; $i < 1000; $i++) {
>             echo ' ';
>         }
> 
>         flush();
>         // Terminate the request
>         if (function_exists('fastcgi_finish_request'))
>             fastcgi_finish_request();
>     }

But it doesn't work and it still appear the blank screen like before. The export file are not downloaded nor email. Then I try the one you said about change the output buffering setting to "On"

> Also, if that doesn't work, try setting ini_set('output_buffering', 'Off'); to ini_set('output_buffering', 'On');

The result was still the same one, it appears blank screen as before. That's mean it still doesn't work smoothly and there is still a problem regarding this issue. So, what to do next sir? Glad to try your future idea of solution from your upcoming reply!

Thanks

**Comment by JediKev** [comment]
@umisuhanarahman 

You'll have to contact your host then. Tell them you updated the code to look exactly like the example they referenced (from binarytides) and it's still not working. I'm sure they have custom configurations in place preventing this from working.

Cheers.

**Comment by Dav1dNL** [comment]
> I have the exact same issue but a little different, for me this only happens from Agent Panel->Users->Organizations->Choose Any Organization->Tickets->Export. I am also on 1.14.1 upgraded from 1.12.5 (it worked fine in that version). Is there a fix or temporary workaround to this? It is the end of the month and we need to generate customer billing info based on per-organization basis. Pls kindly help! Thanks in advance!

I have exact the same issue as you.
Did you manage to fix the problem and can you help me?

**Comment by JediKev** [comment]
@DavidCommandeur 

See the following comment for a patch:
- https://github.com/osTicket/osTicket/issues/5303#issuecomment-589091466

Cheers.

**Comment by Dav1dNL** [comment]
Hi @JediKev 

Tried it but still the white bar. I'am getting keen on fixing this problem.

I find it strange that the export problem only occures when making an export in the scp:
'user tab -> user directory ->user -> tickets tab -> export' 
and: 
'user tab -> organizations -> organisation -> ticket tab -> export'

All other exports and even the 'user tab -> user directory -> export' I can make an export without a problem.

what are the differences?

for your information, see thread: https://forum.osticket.com/d/96770-organization-ticket-export-not-working/2

**Comment by JediKev** [comment]
@DavidCommandeur 

Okay so your issue is different than the OP’s issue. OP’s issue is due to Shared Host restrictions. Your issue deals with the organization export specifically (which is due to undefined method) which I’m currently working on. A fix will be posted to the thread you linked on the Forum when I’m done.

Cheers.

**Comment by Dav1dNL** [comment]
@JediKev 

Yes that's the case.
Ik cant't export the tickets from user or organization under the user tab.

The list of users or the list of organizations itself I can export, like all other exports.

thanks in advance for the upcomming fix.

**Comment by gtao725** [comment]
@JediKev @Dav1dNL https://github.com/osTicket/osTicket/pull/5418/files this has also fixed my issue, thx so much guys!!!

@umisuhanarahman Hope everything works out for you too.

**Comment by Lamboy88** [comment]
I do still have the problem. following the instrunctions I get the error in php.log

[01-Dec-2020 20:49:40 UTC] PHP Warning:  Cannot modify header information - headers already sent by (output started at /xx/include/class.http.php:86) in /xx/include/class.http.php on line 94
[01-Dec-2020 20:49:40 UTC] PHP Warning:  Cannot modify header information - headers already sent by (output started at /xx/include/class.http.php:86) in /xx/include/class.http.php on line 95

from the lines it is referring to the following:
(94)        header('Content-Type: text/plain');
(95)        header('Cache-Control: no-cache');

when commenting out this does disappear and no errors are shown.
Indiferrently of ini_set('output_buffering', 'Off'); or On the csv is only send by email. 

Following the advices above: commenting out the email send, does not help, still nothing is shown.

I have included into htaccess
php_value output_buffering On
php_admin_value upload_tmp_dir /xx/tmp (for testing 777)

![grafik](https://user-images.githubusercontent.com/16718088/100796861-0bf23880-3400-11eb-97c4-e7a47a1878fc.png)


Any idea what could be the problem?
thank you!

**Comment by JediKev** [comment]
@Lamboy88 

Are you using a shared host or hosting provider? Or is this a server hosted in-house?

Cheers.

**Comment by Lamboy88** [comment]
shared host.

**Comment by JediKev** [comment]
@Lamboy88 

`v1.14.3` should already include the following patch but please check to make sure:
- https://github.com/osTicket/osTicket/pull/5418

If you already have that patch applied, please follow the instructions listed [here](https://github.com/osTicket/osTicket/issues/5303#issuecomment-588310784) and if that doesn't work follow the instructions listed [here](https://github.com/osTicket/osTicket/issues/5303#issuecomment-589091466). If it still doesn't work you need to contact your host to see how to get around their specific restrictions.

Cheers.

**Comment by Lamboy88** [comment]
Thank you @JediKev unfortunately everything already tested without getting it fixed. Will try to get some ideas on what could be wrong here. cheers

**Comment by JediKev** [comment]
@Lamboy88 

Yea, your host definitely has some weird, custom restrictions. Contacting them would be in your best interest.

Cheers.

**Comment by juliogaraujo** [comment]
Hello All,

I'm facing exactly the same problem and I would like to know if there is any update about this issue?

Thanks,
Julio

**Comment by JediKev** [comment]
@juliogaraujo 

I would encourage you to read [this comment](https://github.com/osTicket/osTicket/issues/5303#issuecomment-736830519) above and try the steps listed.

Cheers.

**Comment by juliogaraujo** [comment]
@JediKev 

Thanks for your prompt reply and sorry for not mention before that I read all this post and I tried everything that is suggested here and in the linked posted and the only thing that works was comment the lines at ajax.tickets.php file:
//$exporter->email($thisstaff);
//@$exporter->delete();

But seems that commenting this lines will cause a side effect that the email is never send, so I'm looking for a more better workaround or even a final solution for this problem.

Is there any other solution?

Thanks once again.
Cheers.

**Comment by JediKev** [comment]
@juliogaraujo 

Are you using a shared host? If so, I’d recommend contacting your hosting provider for further assistance.

Cheers.

**Comment by juliogaraujo** [comment]
@JediKev 

Yep sir. I'm using a shared host and I already tried to contact them but they couldn't figure out find the problem. 

Do you mid if you can tell me what exactly I should ask to them?

Also I build a environment inside the customer network using webmin and I got exactly the same issue, anyway for this environment I have full access, so is there any test that I can do?

Cheers.

**Comment by JediKev** [comment]
@juliogaraujo 

I would [start here](https://github.com/osTicket/osTicket/issues/5303#issuecomment-585807641) and make sure you’ve tried everything listed in that comment and all 7 comments directly following it.

It has honestly been a while since I’ve touched on this issue but I explained in [this comment](https://github.com/osTicket/osTicket/issues/5303#issuecomment-585410716) what I found when using a shared host environment that was experiencing the same issue. It was due to Output Buffering that the host intentionally disabled so I explained what I had to do to get it working in [this comment](https://github.com/osTicket/osTicket/issues/5303#issuecomment-589091466).

Cheers.

**Comment by juliogaraujo** [comment]
@JediKev 

Thanks a lot for your comments, I will double check everything from the scratch to make sure that I didn't missed anything.

Cheers.

**Comment by RikkiMasters** [comment]
Same issue here after updating to 1.15.2 from 1.12.2.

Self hosted on IIS
PHP 7.3 
Output buffering on

Commenting out the below fixes the issue.

> include/ajax.tickets.php:
> 
> //$exporter->email($thisstaff);
> and
> //@$exporter->delete();

Have tried this fix but no luck: https://github.com/osTicket/osTicket/issues/5303#issuecomment-589091466 


![image](https://user-images.githubusercontent.com/11843864/122745678-25b27a80-d281-11eb-8eb2-1b44433c4433.png)

EDIT: Have tested on different PHP versions. It works on 7.2, but fails on 7.3. and 7.4.

Downgrading to 7.2 fixes the issue. Whether it's because my config is different in 7.2 I don't know. Any specifics I can check for?

**Comment by ntozier** [comment]
Server 2019 and IIS 10, PHP 7.3.19, osT 1.15.2.
Same issue.  Tries steps to resolve above with no luck.
Additionally tried setting responseBufferLimit to 0 in the web.config.

```
<handlers>
<remove name="FastCGI" />
<add name="FastCGI" path="*.php" verb="*" type="" modules="FastCgiModule" scriptProcessor="D:\PHP\7_3_19\php-cgi.exe" resourceType="File" requireAccess="Script" allowPathInfo="false" preCondition="" responseBufferLimit="0" />
</handlers> 
```

For more information regarding this attempt see:
https://stackoverflow.com/questions/7178514/php-flush-stopped-flushing-in-iis7-5

**Comment by RikkiMasters** [comment]
@ntozier did you try PHP 7.2?

**Comment by ntozier** [comment]
@RikkiMasters no I did not.  PHP 7.2 is end of life.  I'm more apt to go to 7.4 (or 8 once osTicket supports it).

**Comment by RikkiMasters** [comment]
@ntozier My point is it works on 7.2. So testing it would narrow down the issue to a 7.3 change.

**Comment by ghostersk** [comment]
we are using Plesk to host this ticket system. I change 2 php files as per this post 
https://github.com/osTicket/osTicket/pull/5418/commits/0cd4168e5d6e428ff85448f61cd28d930e198716
and then I added TMP folder link to PHP settings :
sys_temp_dir = "/tmp"
upload_tmp_dir = "/tmp"

Also I am using PHP 7.3
and I was able to export Tickets.

**Comment by apyash** [comment]
I am also facing the same issue when newly installed with below configured. Task, Org user export working, but ticket export is not working.
--
Version | v1.15.4 (6bd7884) —  Up to date
Web Server Software | Apache
MySQL Version | 5.6.41
PHP Version | 7.4.26

![image](https://user-images.githubusercontent.com/49680362/143575683-176557f7-3743-4121-9919-2c437c2190fb.png)

**Comment by JediKev** [comment]
@apyash 

I am not able to replicate this issue using v1.15.4, Apache, MySQL 8, and PHP 7.4. Please review all logs on your server for any related errors (general server logs, webserver error logs, PHP error logs, MySQL/MariaDB error logs, osTicket System Logs, Browser Console logs, etc.). The empty white box is typically indicative of either AJAX issues or a fatal PHP error that should be logged somewhere. I think it's the latter as this is most common with export issues.

Cheers.

**Comment by apyash** [comment]
In logs we can see only
[Sun Nov 28 22:39:31.836045 2021] [:error]  File does not exist: xmlrpc.php
[Mon Nov 29 20:39:05.746064 2021] [:error]] File does not exist: xmlrpc.php

Also no mod security errors found.

As per Hosted Server Message: Looks like coding issue as per their message
I have installed the new version as fresh, but same issue "xmlrpc.php files itself not found in the account" as per them.

**Comment by jmm312** [comment]
Same issue

Server 2019 
Microsoft-IIS/10.0
PHP 8.0.16
osTicket v1.16.1
output_buffering = 4096 (I have also tried changing this to On)
Write permissions have been granted to the Temp folder. Confirmed a temp file is being created.

A short-term fix that worked was commenting out the following sections in the ajax.tickets.php file:

>     include/ajax.tickets.php:
> 
>     //$exporter->email($thisstaff);
>     and
>     //@$exporter->delete();

**Comment by marcosarossi** [comment]
I think I found the problem!. It has something to do with the session storage between requests. I tried exporting tickets using a browser speed regulation (thus making the requests slower) and worked fine. That gave me a hint.

So, in order to fix this, I edited the **class.export.php**, function **ack()**. This is my version:

```php
function ack() {
        // Register the export in the session
        self::register($this);

        // IT GOES HERE
        session_write_close();

        // Flush response / return export id and check interval
        Http::flush(201, json_encode(['eid' =>
                    $this->getId(), 'interval' => $this->getInterval()]));
        // Phew... now we're free to do the export
        // NOT HERE!: session_write_close(); // Release session for other requests
        ignore_user_abort(1);  // Leave us alone bro!
        @set_time_limit(0);    // Useless when safe_mode is on
        // Ask the queue to export to the exporter
    }
```

Hope it helps!
